pkg/apiserver/rest/dualwriter.go

```diff
@@ -73,36 +87,98 @@ func NewDualWriter(legacy LegacyStorage, storage Storage) *DualWriter {
 	}
 }
 
-// Create overrides the default behavior of the Storage and writes to both the LegacyStorage and Storage.
+// Get overrides the default behavior of the Storage and retrieves an object from
+// LegacyStorage or Storage depending on the DualWriter mode.
+func (d *DualWriter) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	legacy, ok := d.legacy.(rest.Getter)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Getter is missing")
+	}
+
+	switch CurrentMode {
+	case Mode1, Mode2:
+		return legacy.Get(ctx, name, &metav1.GetOptions{})
+	case Mode3, Mode4:
+		return d.Storage.Get(ctx, name, &metav1.GetOptions{})
+	}
+
+	return nil, fmt.Errorf("dual writer mode is invalid")
+}
+
+// Create overrides the default behavior of the Storage and writes to LegacyStorage and Storage depending on the dual writer mode.
 func (d *DualWriter) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
-	if legacy, ok := d.legacy.(rest.Creater); ok {
-		created, err := legacy.Create(ctx, obj, createValidation, options)
-		if err != nil {
-			return nil, err
+	legacy, ok := d.legacy.(rest.Creater)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Creater is missing")
+	}
+
+	created, err := legacy.Create(ctx, obj, createValidation, options)
+	if err != nil {
+		return nil, err
+	}
+
+	if CurrentMode < Mode2 {
+		return created, nil
+	}
+
+	rsp, err := d.Storage.Create(ctx, created, createValidation, options)
+	// #TODO customise error handling depending on the mode we are in
+	if err != nil {
+		klog.Error("unable to create object in duplicate storage", "error", err)
+	}
+	return rsp, err
+}
+
+// Update overrides the default behavior of the Storage and writes to both the LegacyStorage and Storage depending on the DualWriter mode.
+func (d *DualWriter) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
+	// #TODO: figure out which method should be used anytime Get is called
+	// #TODO figure out failure modes, how to guarantee consistency of the updates across both stores
+	// #TODO replace with a rest.CreaterUpdater check?
+	legacy, ok := d.legacy.(rest.Updater)
+	if !ok {
+		return nil, false, fmt.Errorf("legacy storage rest.Updater is missing")
+	}
+
+	// #TODO: Doing it the repetitive way first. Refactor once all the behavior is well defined.
```

Comment:
An alternative that may be worth exploring would be 4 different dual writer implementations, 1 for each mode (with mode 1 and mode 4 just being passthrough to legacy and unified respectively), so that we can diff between mode 2 and 3 easily

Reply:
True. One more alternative would be adding mode as a DualWriter field. That's still quite close to the current implementation though. 

I'll check out the alternatives before finalising the PR.